### PR TITLE
multi: Log swap refund txs.

### LINKED ISSUE #1087: log a full signed refund txn when sending a swap
https://github.com/decred/dcrdex/blob/3fa59cf4dad60a991346795da25666f9bd215af2/client/core/trade.go#L1399

At this point, I believe the bytes of the signed refund transaction should be logged for emergency use.  My suggestion is that the `client.Asset.Swap` method also return these bytes.

Related: https://github.com/decred/dcrdex/issues/952

### PR DESCRIPTION
**Pull request by JoeGruffins**
Add signed refund txs to the Receipt when swapping and log so that the
    user can salvage funds in the case that they lose access to the client
    but still have the logs.

closes #1087

### CODE REVIEW

**Review by chappjc — COMMENTED** [review]
Good stuff, but oh boy did this give the user a hell of a foot-gun, especially with Decred where a non-final txn like the refund can enter mempool before locktime to wait for mining.

bitcoind wouldn't allow broadcast of the refund tx early, just responding with:

```
error code: -26
error message:
non-final
```

But with the Decred refund in mempool, all kinds of stuff broke.  The actual redeem txn couldn't enter mempool, being rejected with `error sending redeem transaction: -4: wallet.PublishTransaction(6f775...): double spend:: wallet.processTransactionRecord: 6f775... conflicts with 47ca3f... by double spending 75ae3a...:0}`.
That obviously breaks the swap and client can't fix it.  It also makes redemption search bomb out endlessly forever.

Looks like we need to consider how to prevent this, and consider how dexc/Core should recover from the myriad of edge cases hit when a refund is broadcast outside of it's control.  :sigh:

**Comment by chappjc** [comment]
> Good stuff, but oh boy did this give the user a hell of a foot-gun, especially with Decred where a non-final txn like the refund can enter mempool before locktime to wait for mining

Ah, I forgot that this is because only on Decred's **mainnet** are non-standard transactions blocked from entering mempool or being relayed!  Only on testnet and simnet should this be a problem.  And it should help use test out the edge cases.

https://github.com/decred/dcrd/blob/2e8efe15a335f654ccacbef23ae8616181ae20a3/internal/mempool/mempool.go#L1307-L1312

https://github.com/decred/dcrd/blob/2e8efe15a335f654ccacbef23ae8616181ae20a3/internal/mempool/policy.go#L315-L319

File: client/asset/dcr/dcr.go [hunk 1]
```diff
@@ -2047,9 +2089,11 @@ func (dcr *ExchangeWallet) Refund(coinID, contract dex.Bytes) (dex.Bytes, error)
 		return nil, fmt.Errorf("refund tx not worth the fees")
 	}
 
-	refundAddr, err := dcr.node.GetNewAddressGapPolicy(dcr.ctx, dcr.acct, dcrwallet.GapPolicyIgnore)
-	if err != nil {
-		return nil, fmt.Errorf("error getting new address from the wallet: %w", translateRPCCancelErr(err))
+	if refundAddr == nil {
+		refundAddr, err = dcr.node.GetNewAddressGapPolicy(dcr.ctx, dcr.acct, dcrwallet.GapPolicyIgnore)
```
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/asset/dcr/dcr.go:2093]: I believe it should be an external address. It is `meant to be visible outside of the wallet` https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#change

[the remark this answers is not in the document]
Comment (chappjc) [reply client/asset/dcr/dcr.go:2093]: It's not related too this PR for sure; it just popped out.  Although TBH, I don't think I get the wording "meant to be visible outside of the wallet" except for their example "for receiving payments".  So this is walking the line of semantics of "for receiving payments" (external) vs. "return transaction change" (internal).  The revoke address used inside the contract script that you yourself have authored is a bit like smart contract version of change. And even the refund transaction's output is in fact paying to yourself from an output (the refunded contract) that you also technically own.  I get the feeling bip 44 wasn't written with these use cases in mind.

File: client/core/trade.go [hunk 2]
```diff
@@ -1466,8 +1466,16 @@ func (c *Core) swapMatchGroup(t *trackedTrade, matches []*matchTracker, errs *er
...
+	refundTxs := ""
+	for i, r := range receipts {
+		refundTxs = fmt.Sprintf("%s%q: %s", refundTxs, r.Coin(), r.SignedRefund())
+		if i != len(receipts)-1 {
+			refundTxs = fmt.Sprintf("%s, ", refundTxs)
+		}
+	}
+
+	c.log.Infof("Broadcasted transaction with %d swap contracts for order %v. Fee rate = %d. Receipts (%s): %v. Refund Txs: {%s}",
+		len(receipts), t.ID(), swaps.FeeRate, t.wallets.fromAsset.Symbol, receipts, refundTxs)
```
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/core/trade.go:1478]: Does this look ok?
```
2021-07-07 12:11:15.035 [INF] CORE: The following are coin id mapped to raw refund transactions that are only valid after the swap contract expires. They can be used to return funds to your wallet in the case dexc no longer functions. They SHOULD NOT be used if dexc is running without error. dexc will refund failed contracts automatically.
Refund Txs: {"e680826a7a11754bdc9744159ce41c7b26cc528287065ecf1c58978caeff7a82:0": 01000000000101827affae8c97581ccf5e06878252cc267b1ce49c154497dc4b75117a6a8280e60000000000feffffff01f2c89a3b00000000160014e7608cf1b0d005b3119e708fce32d2bb21a511e804483045022100fa9cbdc1a113776fb46e4969076d7471a056eb2b6b09c1f2dca4f92b441156c0022044beca578f81fd23cd5d1696d38082c281f4fae46435c4f2f0eb6cdc2c4b5ad7012103d5da2fdce9120541829344d5efa485eba3188773242c0a50ecfa9ca4217e1b9200616382012088a82093321e6493eac955c7c5c6bda721dbc1b7afbf6c319fb5cedfff6ba8d0eb8e0f8876a914100dc814d156d20cd673ce2936692aaf5ef9f2b567049334e660b17576a914e7608cf1b0d005b3119e708fce32d2bb21a511e86888ac9334e660}
```

File: client/asset/btc/btc.go [hunk 3]
```diff
@@ -1352,6 +1361,7 @@ func (btc *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin
 		if err != nil {
 			return nil, nil, 0, fmt.Errorf("error creating revocation address: %w", err)
 		}
+		refundAddrs = append(refundAddrs, revokeAddr)
```
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/asset/btc/btc.go:1364]: `revokeAddr is the address belonging to the key that will be used to sign and refund a swap past its encoded refund locktime.`

File: client/core/trade.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (chappjc) [reply client/core/trade.go:1478]: Yeah, pretty clear except maybe change the phrase "coin id mapped to raw refund transactions" to "contract identifiers mapped to raw refund transactions" (or similar).  Both "coin id" and "contract identifiers" are obv generic ways of saying the the tx output in utxo land, but I don't think anyone but dex devs will know what to make of "coin id".

Also maybe "They can be used" -> "These are fallback transactions that can be used"... just to be extra clear that they don't have to panic copy/paste/send it or anything.

[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/core/trade.go]: Will do.

[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/core/trade.go]: Made those changes. Hopefully reads ok.

**Review by chappjc — APPROVED** [review]
Looking good and works well.  `TestNoMakerRedeem` (the one with two refunds) working still.
